README.md
# finance_rb

This package is a ruby native port of the numpy-financial package with some helpful additional functions.

The functions in this package are a scalar version of their vectorised counterparts in the [numpy-financial](https://github.com/numpy/numpy-financial) library.

![tests](https://github.com/wowinter13/finance_rb/actions/workflows/tests.yml/badge.svg) [![Release](https://img.shields.io/github/v/release/wowinter13/finance_rb.svg?style=flat-square)](https://github.com/wowinter13/finance_rb/releases) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com) [![Maintainability](https://api.codeclimate.com/v1/badges/bbca82ad7815794c6718/maintainability)](https://codeclimate.com/github/wowinter13/finance_rb/maintainability)

Currently, only some functions are ported,  
which are as follows:  

| numpy-financial function     | ruby native function ported?   | info|
|:------------------------:    |:------------------:  | :------------------|
| fv                           |   ✅    |   Computes the  future value|
| ipmt                         |   ✅   |   Computes interest payment for a loan|
| pmt                          |  ✅    |   Computes the fixed periodic payment(principal + interest) made against a loan amount|
| ppmt                         |   ✅   |   Computes principal payment for a loan|
| nper                         |    ✅   |    Computes the number of periodic payments|
| pv                           |    ✅  |   Computes the present value of a payment|
| rate                         |   ✅ |    Computes the rate of interest per period|
| irr                          |    ✅    |    Computes the internal rate of return|
| npv                          |  ✅   |   Computes the net present value of a series of cash flow|
| mirr                         |    ✅    |    Computes the modified internal rate of return|


**Things to be done:**

1. Xirr
2. More specs for edge cases
3. Fee, currency protection and other cool stuff for advanced usage
4. Better errors
5. A detailed documentation for all methods

## Documentation

Detailed documentation is available at [rubydoc](https://rubydoc.info/gems/finance_rb).

### Basic usage

There is no need to create a loan class to calculate some basic functions, such as `npv`, `irr`, or `mirr`. Simply call a needed method with required arguments using calculations class.

1. **IRR** (Internal Rate of Return)

    ```ruby
    Finance::Calculations.irr([-4000,1200,1410,1875,1050])
    => 0.14299344106053188
    ```

2. **MIRR** (Modified Internal Rate of Return)

    ```ruby
    Finance::Calculations.mirr([100, 200, -50, 300, -200], 0.05, 0.06)
    => 0.3428233878421769
    ```

3. **NPV** (Net Present Value)

    ```ruby
    Finance::Calculations.net_present_value(0.1, [-100, 6, 6, 6])
    => -85.07888805409468
    ```

### Advanced Usage

Create a loan instance, and pass available parameters for nominal annual rate, duration (in months), and amount of loan.

```ruby

loan = Finance::Loan.new(nominal_rate: 0.1, duration: 12, amount: 1000)
loan.pmt # => -87.9158872300099
```

Available methods: `pmt`, `ipmt`, `ppmt`, `nper`, `pv`, `fv`, `rate`.

## Installation

finance_rb is available as a gem, to install it just install the gem:

    gem install finance_rb

If you're using Bundler, add the gem to Gemfile.

    gem 'finance_rb'

Run `bundle install`.

## Running tests

    bundle exec rspec spec/

## Contributing

1. Fork it ( https://github.com/wowinter13/finance_rb/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## License

MIT License. See LICENSE for details.

CHANGELOG.md
## [1.0.0] - 2021-05-01

### Added
* Implement `Finance::Loan#rate`

## [0.2.2] - 2021-05-01

### Added
* Implement `Finance::Loan#pv`

## [0.2.1] - 2021-05-01

### Added
* Implement `Finance::Loan#nper`

## [0.2.0] - 2021-04-30

### Added
* Implement `Finance::Loan#ppmt`

### Changed
* By default `Finance::Loan#pmt` returns negative values.

## [0.1.2] - 2021-04-05

### Added
* Implement `Finance::Loan#ipmt`

## [0.1.1] - 2021-03-30

### Added
* Implement `Finance::Loan#fv`

## [0.1.0] - 2021-03-28

### Added
* Create a basic structure for `Finance::Loan`
* Implement `Finance::Loan#pmt`


## [0.0.4] - 2021-03-25

### Added
* Implement `Finance::Calculations#mirr`

## [0.0.3] - 2021-03-23

### Added
* Implement `Finance::Calculations#irr` using Newton's method

## [0.0.2] - 2021-03-22

- Initial public release


## [0.0.1] - 2021-03-20

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true

require "finance/calculations"
require "finance/loan"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/finance/rb_spec.rb
# frozen_string_literal: true

RSpec.describe Finance::Rb do
  it "has a version number" do
    expect(Finance::Rb::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/finance/loan_spec.rb
require 'spec_helper'

RSpec.describe Finance::Loan do
  describe '#pmt' do
    context 'w/o a full set of params' do
      it 'calculates correct pmt value w/o :ptype' do
        loan = Finance::Loan.new(nominal_rate: 0.1, duration: 12, amount: 1000)
        expect(loan.pmt).to eq(-87.9158872300099)
      end

      it 'calculates correct pmt value w/o :nominal_rate' do
        loan = Finance::Loan.new(duration: 12, amount: 1200, ptype: :end)
        expect(loan.pmt).to eq(-100)
      end
    end

    context 'with zero rates' do
      it 'calculates correct pmt value for 3 years' do
        loan = Finance::Loan.new(nominal_rate: 0, duration: 36, amount: 10_000, ptype: :end)
        expect(loan.pmt).to eq(-277.77777777777777)
      end

      it 'calculates correct pmt value for 6 months' do
        loan = Finance::Loan.new(nominal_rate: 0, duration: 6, amount: 10_000, ptype: :end)
        expect(loan.pmt).to eq(-1666.6666666666667)
      end
    end

    context 'with :beginning ptype' do
      it 'calculates correct pmt value' do
        loan = Finance::Loan.new(nominal_rate: 0.12, duration: 6, amount: 1000, ptype: :beginning)
        expect(loan.pmt).to eq(-170.8399670404763)
      end
    end

    it 'calculates correct pmt value' do
      loan = Finance::Loan.new(nominal_rate: 0.13, duration: 90, amount: 1_000_000, ptype: :end)
      expect(loan.pmt).to eq(-17_449.90775727763)
    end
  end

  describe '#fv' do
    context 'with loan arguments' do
      it 'calculates correct fv value' do
        loan = Finance::Loan.new(nominal_rate: 0.05, duration: 120, amount: -100, payment: -100)
        expect(loan.fv).to eq(15_692.928894335748)
      end

      context 'with :ptype' do
        it 'calculates correct fv value' do
          loan = Finance::Loan.new(
            nominal_rate: 0.9, duration: 20, amount: 0, payment: -2000, ptype: :beginning
          )
          expect(loan.fv).to eq(93_105.06487352113)
        end
      end
    end

    context 'with an optional :payment argument' do
      it 'calculates correct fv value' do
        loan = Finance::Loan.new(nominal_rate: 0.05, duration: 120, amount: -100, payment: -200)
        expect(loan.fv(payment: -100)).to eq(15_692.928894335748)
      end
    end

    context 'w/o any payments' do
      it 'raises an ArgumentError exception w/o loan arguments' do
        loan = Finance::Loan.new(nominal_rate: 0.05, duration: 120, amount: -100)
        expect { loan.fv }.to raise_error(ArgumentError, "no payment given")
      end
    end
  end

  describe '#pv' do
    context 'when :ptype == beginning' do
      it 'calculates correct pv value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.24, duration: 12, future_value: 1000, payment: -300, ptype: :beginning
        )
        expect(loan.pv).to eq(2447.5612380190028)
      end
    end

    context 'when :ptype == ending' do
      it 'calculates correct pv value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.24, duration: 12, future_value: 1000, payment: -300, ptype: :ending
        )
        expect(loan.pv).to eq(2384.1091906935)
      end
    end
  end

  describe '#ipmt' do
    context 'when 1 period' do
      it 'calculates correct ipmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 1
        )
        expect(loan.ipmt).to eq(-17.166666666666668)
      end
    end

    context 'when 2 periods' do
      it 'calculates correct ipmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 2
        )
        expect(loan.ipmt).to eq(-15.789337457350777)
      end
    end

    context 'when 3 periods' do
      it 'calculates correct ipmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12.0, amount: 2500.0, period: 3.0, fv: 0.0
        )
        expect(loan.ipmt).to eq(-14.402550587464257)
      end
    end
  end

  describe '#ppmt' do
    context 'when 1 period' do
      it 'calculates correct ppmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 1
        )
        expect(loan.ppmt).to eq(-200.58192368678277)
      end
    end

    context 'when 2 periods' do
      it 'calculates correct ppmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 2
        )
        expect(loan.ppmt).to eq(-201.95925289609866)
      end
    end

    context 'when 3 periods' do
      it 'calculates correct ppmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 3
        )
        expect(loan.ppmt).to eq(-203.34603976598518)
      end
    end

    context 'when 4 periods' do
      it 'calculates correct ppmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 4
        )
        expect(loan.ppmt).to eq(-204.7423492390449)
      end
    end

    context 'when 5 periods' do
      it 'calculates correct ppmt value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.0824, duration: 12, amount: 2500, period: 5
        )
        expect(loan.ppmt).to eq(-206.1482467038197)
      end
    end
  end

  describe '#nper' do
    context 'with normal arguments' do
      it 'calculates correct nper value' do
        loan = Finance::Loan.new(
          nominal_rate: 0.07, amount: 8000, payment: -150, future_value: 0 
        )
        expect(loan.nper).to eq(64.0733487706618586)
      end
    end

    context 'with incorrect arguments' do
      it 'raises Math::DomainError' do
        loan = Finance::Loan.new(
          nominal_rate: 1e100, amount: 8000, payment: -150, future_value: 0 
        )
        expect { loan.nper }.to raise_error(Math::DomainError)
      end
    end
  end

  describe '#rate' do
    context 'with default arguments' do
      it 'calculates correct rate value' do
        loan = Finance::Loan.new(
          nominal_rate: 10, amount: -3500, payment: 0, duration: 10, future_value: 10000
        )
        expect(loan.rate).to eq(0.11069085371426901)
      end
    end
  end
end


lib/finance/calculations.rb
# frozen_string_literal: true

require 'bigdecimal'
require 'bigdecimal/newton'

include Newton

module Finance
  class Calculations
    class << self
      # Npv computes the Net Present Value of a cash flow series.
      #
      # @return [Numeric] The NPV of the input cash flow series `values` at the discount `rate`.
      #
      # @param [Numeric] :rate A discount rate applied once per period.
      # @param [Array<Numeric>] :values The values of the time series of cash flows.
      # 
      # @example
      #   require 'finance_rb'
      #   Finance::Calculations.npv(0.1, [-1000, 100, 100, 100]) #=> -789.3518518518517
      #
      # @see http://en.wikipedia.org/wiki/Net_present_value
      # @see L. J. Gitman, “Principles of Managerial Finance, Brief,” 3rd ed.,
      #   Addison-Wesley, 2003, pg. 346.
      def npv(rate, values)
        npv_value = 0.0
        values.each_with_index do |current_value, pow_index|
          npv_value += current_value / (1.0 + rate) ** pow_index
        end
        npv_value
      end

      # IRR computes the Rate of Interest per period.
      #
      # @return [Float] Internal Rate of Return for periodic input values.
      #
      # @param [Array<Numeric>] :values Input cash flows per time period.
      #   At least, must contain one positive and one negative value.
      #   Otherwise, irr equals zero.
      #
      # @example
      #   require 'finance_rb'
      #   Finance::Calculations.irr([-100, 0, 0, 74]) #=> 0.14299344106053188
      #
      # @see http://en.wikipedia.org/wiki/Internal_rate_of_return
      # @see L. J. Gitman, "Principles of Managerial Finance, Brief," 3rd ed.,
      #   Addison-Wesley, 2003, pg. 348.
      def irr(values)
        return 0.0 unless correct_cashflows?(values)

        func = BigDecimal.limit(100)
        func = Function.new(values)
        rate = [ func.one ]
        nlsolve(func, rate)
        rate[0].to_f
      end

      # MIRR computes the modified Rate of Interest.
      #
      # @return [Float] Modified Internal Rate of Return.
      #
      # @param [Array<Numeric>] :values
      #   At least, must contain one positive and one negative value.
      #   Otherwise, mirr equals zero.
      # @param [Numeric] :rate Interest rate paid on the cash flows
      # @param [Numeric] :reinvest_rate Interest rate received on the cash flows upon reinvestment
      # 
      # @example
      #   require 'finance_rb'
      #   Finance::Calculations.mirr([100, 200, -50, 300, -200], 0.05, 0.06) => 0.2979256979689131
      #
      # @see https://en.wikipedia.org/wiki/Modified_internal_rate_of_return
      def mirr(values, rate, reinvest_rate)
        inflows = [];
        outflows = [];
        # We prefer manual enumeration over the partition
        #   because of the need to replace outflows with zeros.
        values.each do |val|
          if val >= 0
            inflows << val
            outflows << 0.0
          else
            outflows << val
            inflows << 0.0
          end
        end
        if outflows.all?(0.0) || inflows.all?(0.0)
          return 0.0
        end
        fv = npv(reinvest_rate, inflows).abs
        pv = npv(rate, outflows).abs

        return (fv/pv) ** (1.0/(values.size - 1)) * (1 + reinvest_rate) - 1
      end

      alias net_present_value npv
      alias internal_return_rate irr

      private

      # @api private
      def correct_cashflows?(values)
        inflows, outflows = values.partition{ |i| i >= 0 }
        !(inflows.empty? || outflows.empty?)
      end

      # Base class for working with Newton's Method.
      # For more details, see Bigdecimal::Newton.
      #
      # @api private
      class Function
        def initialize(values)
          @zero   = BigDecimal("0.0")
          @one    = BigDecimal("1.0")
          @two    = BigDecimal("2.0")
          @ten    = BigDecimal("10.0")
          @eps    = BigDecimal("1.0e-16")
          @values = values
        end
        
        def zero; @zero; end
        def one ; @one;  end
        def two ; @two;  end
        def ten ; @ten;  end
        def eps ; @eps;  end

        def values(x); [Finance::Calculations.npv(x[0], @values)]; end
      end
    end
  end
end


lib/finance/loan.rb
# frozen_string_literal: true

module Finance
  class Loan
    PAYMENT_TYPE_MAPPING = { end: 0.0, beginning: 1.0 }.freeze

    # @return [Float] The amount of loan request (I.e. a present value)
    #   You can use #pv method to calculate value if param is not defined.
    #   Defaults to 0.
    attr_accessor :amount

    # @return [Integer] Specification of whether payment is made
    #   at the beginning (ptype = 1) or the end (ptype = 0) of each period.
    #   Defaults to {:end, 0}.
    attr_accessor :ptype

    # @return [Float] The nominal annual rate of interest as decimal (not per cent).
    #   (e.g., 13% -> 0.13)
    #   Defaults to 0.
    attr_accessor :nominal_rate

    # @return [Float] The monthly rate is the nominal annual rate divided by 12.
    #   Defaults to 0.
    attr_reader :monthly_rate

    # @return [Float] The number of periods to be compounded for. (I.e. Nper())
    #   Defaults to 1.
    #   You can use #nper method to calculate value if param is not defined.
    attr_accessor :duration

    # @return [Float] Future value.
    #   You can use #fv method to calculate value if param is not defined.
    #   Defaults to 0.
    attr_accessor :future_value

    # @return [Float] The (fixed) periodic payment.
    #   You can use #pmt method to calculate value if param is not defined.
    attr_accessor :payment

    # @return [Float] Period under consideration.
    attr_accessor :period

    # Create a new Loan instance.
    def initialize(**options)
      initialize_payment_type(options[:ptype])
      @nominal_rate  = options.fetch(:nominal_rate, 0).to_f
      @duration      = options.fetch(:duration, 1).to_f
      @amount        = options.fetch(:amount, 0).to_f
      @future_value  = options.fetch(:future_value, 0).to_f
      @period        = options[:period]
      @payment       = options[:payment]
      @monthly_rate  = @nominal_rate / 12
    end

    # Pmt computes the payment against a loan principal plus interest (future_value = 0).
    #   It can also be used to calculate the recurring payments needed to achieve
    #   a certain future value given an initial deposit,
    #   a fixed periodically compounded interest rate, and the total number of periods.
    #
    #   Required Loan arguments: nominal_rate, duration, amount, future_value*
    #
    # @return [Numeric] The (fixed) periodic payment.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.1, duration: 12, amount: 1000, ptype: :end).pmt
    #   #=> 87.9158872300099
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def pmt
      factor = (1.0 + monthly_rate)**duration
      second_factor =
        if monthly_rate.zero?
          duration
        else
          (factor - 1) * (1 + monthly_rate * ptype) / monthly_rate
        end

      -((future_value + amount * factor) / second_factor)
    end

    # IPmt computes interest payment for a loan under a given period.
    #
    #   Required Loan arguments: period, nominal_rate, duration, amount, future_value*
    #
    # @return [Float] The interest payment for a loan.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.0824, duration: 12, amount: 2500, period: 1).ipmt
    #   #=> -17.166666666666668
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def ipmt
      raise ArgumentError, 'no period given' if period.nil?

      ipmt_val = remaining_balance * monthly_rate
      if ptype == PAYMENT_TYPE_MAPPING[:beginning]
        period == 1 ? 0.0 : (ipmt_val / 1 + monthly_rate)
      else
        ipmt_val
      end
    end

    # PPmt computes principal payment for a loan under a given period.
    #
    #   Required Loan arguments: period, nominal_rate, duration, amount, future_value*
    #
    # @return [Float] The principal payment for a loan under a given period.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.0824, duration: 12, amount: 2500, period: 1).ppmt
    #   #=> -200.58192368678277
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def ppmt
      pmt - ipmt
    end

    # Nper computes the number of periodic payments.
    #
    #   Required Loan arguments: payment, nominal_rate, period, amount, future_value*
    #
    # @return [Float] The number of periodic payments.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.07, amount: 8000, payment: -150).nper
    #   #=> 64.0733487706618586
    def nper
      z = payment * (1.0 + monthly_rate * ptype) / monthly_rate

      Math.log(-future_value + z / (amount + z)) / Math.log(1.0 + monthly_rate)
    end

    # Fv computes future value at the end of some periods (duration).
    #   Required Loan arguments: nominal_rate, duration, payment, amount*
    #
    # @param payment [Float] The (fixed) periodic payment.
    #   In case you don't want to modify the original loan, use this parameter to recalculate fv.
    #
    # @return [Float] The value at the end of the `duration` periods.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.05, duration: 120, amount: -100, payment: -200).fv
    #   #=> 15692.928894335748
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def fv(payment: nil)
      raise ArgumentError, 'no payment given' if self.payment.nil? && payment.nil?

      final_payment = payment || self.payment

      factor = (1.0 + monthly_rate)**duration
      second_factor = (factor - 1) * (1 + monthly_rate * ptype) / monthly_rate

      -((amount * factor) + (final_payment.to_f * second_factor))
    end

    # Pv computes present value.
    #   Required Loan arguments: nominal_rate, duration, payment, future_value, *ptype
    #
    # @return [Float] The present value.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 0.24, duration: 12, future_value: 1000, payment: -300, ptype: :ending).pv
    #   #=> 2384.1091906935
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def pv
      factor = (1.0 + monthly_rate)**duration
      second_factor = (factor - 1) * (1 + monthly_rate * ptype) / monthly_rate

      -(future_value + (payment.to_f * second_factor)) / factor
    end

    # Rate computes the interest rate per period
    #   by running Newton Rapson to find an approximate value.
    #
    # @return [Float] The interest rate.
    #
    # @param tolerance [Float] Required tolerance for the solution.
    # @param initial_guess [Float] Starting guess for solving the rate of interest.
    # @param iterations [Integer] Maximum iterations in finding the solution.
    #
    # @example
    #   require 'finance_rb'
    #   Finance::Loan.new(nominal_rate: 10, amount: -3500, payment: 0, duration: 10, future_value: 10000).rate
    #   #=> 0.11069085371426901
    #
    # @see http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
    # @see [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
    #   Open Document Format for Office Applications (OpenDocument)v1.2,
    #   Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
    #   Pre-Draft 12. Organization for the Advancement of Structured Information
    #   Standards (OASIS). Billerica, MA, USA. [ODT Document].
    def rate(tolerance: 1e-6, iterations: 100, initial_guess: 0.1)
      next_iteration_rate = nil
      current_iteration_rate = initial_guess

      (0..iterations).each do |iteration|
        next_iteration_rate = current_iteration_rate - rate_ratio(current_iteration_rate)
        break if (next_iteration_rate - current_iteration_rate).abs <= tolerance
        current_iteration_rate = next_iteration_rate
      end

      next_iteration_rate
    end

    private

    # rate_ratio computes the ratio
    #   that is used to find a single value that sets the non-liner equation to zero.
    #
    # @api private
    def rate_ratio(rate)
      t1 = (rate+1.0) ** duration
      t2 = (rate+1.0) ** (duration-1.0)
      g = future_value + t1 * amount + payment * (t1 - 1.0) * (rate * ptype + 1.0) / rate
      derivative_g = \
        (duration * t2 * amount)
          - (payment * (t1 - 1.0) * (rate * ptype + 1.0) / (rate ** 2.0))
          + (duration * payment * t2 * (rate * ptype + 1.0) / rate)
          + (payment * (t1 - 1.0) * ptype/rate)
      
      g / derivative_g
    end

    # @api private
    def initialize_payment_type(ptype)
      @ptype =
        if ptype.nil? || !PAYMENT_TYPE_MAPPING.keys.include?(ptype)
          PAYMENT_TYPE_MAPPING[:end]
        else
          PAYMENT_TYPE_MAPPING[ptype]
        end
    end

    # @api private
    def remaining_balance
      self.class.new(
        nominal_rate: nominal_rate.to_f, duration: period - 1.0,
        amount: amount.to_f, ptype: PAYMENT_TYPE_MAPPING.key(ptype)
      ).fv(payment: pmt)
    end
  end
end
